### ISSUE TITLE
Consider changing tagged-types representation

### ISSUE THREAD

**Issue by marcin-rzeznicki** [softwaremill/scala-common#9]
Yours suffers from the following problem:
```scala
scala> class Container[A]
defined class Container

scala> List(new Container[Int @@ Tag1], new Container[Int @@ Tag2])
<console>:19: error: type mismatch;
 found   : Container[Int @@ Tag2]
    (which expands to)  Container[Int with AnyRef{type Tag <: Tag2}]
 required: Container[_ >: Int @@ Tag2 with Int @@ Tag1 <: Int]
    (which expands to)  Container[_ >: Int with AnyRef{type Tag <: Tag2} with Int with AnyRef{type Tag <: Tag1} <: Int]
Note: Int @@ Tag2 <: Int, but class Container is invariant in type A.
You may wish to define A as +A instead. (SLS 4.5)
```

If you used a trait for representing tags (like Shapeless does) - for instance
```scala
  trait Tagged[+U]
  type @@[+T, +U] = T with Tagged[U]
```

it would have worked:

```scala
scala> List(new Container[Int @@ Tag1], new Container[Int @@ Tag2])
res2: List[Container[_ >: Int with tag.Tagged[Tag2] with Int with tag.Tagged[Tag1] <: Int with tag.Tagged[Object]]] = List(Container@5891f54d, Container@1a5bf46c)
```

**Comment by kciesielski** [comment]
Thanks for reporting, looks like we need to use a trait to resolve this issue and #8, I'll prepare a proper fix.

**Comment by kciesielski** [comment]
Resolved by 0a71e35b04ab32be7119e1a936b94f91e9ee6efe

**Comment by marcin-rzeznicki** [comment]
Thank you :+1:
